xapian-maintainer-tools/profiling/profiler.c

```diff
@@ -0,0 +1,171 @@
+/** @file profiler.c
+ * @brief C file implementing IO logging
+ */
+/* Compile the profiler with:
+ * gcc -shared -fPIC -o profiler.so profiler.c -ldl
+ *
+ * To use it with a process, use:
+ * LD_PRELOAD=/path/to/profiler.so /path/to/executable
+ *
+ * Running this library will produce an strace.log file,
+ * which can be fed to strace-analyse to produce a log, by running:
+ * ./strace-analyse strace.log
+ */
+
+#define _GNU_SOURCE
+
+#include <dlfcn.h>
+#include <fcntl.h>
+#include <string.h>
+
+#include <stdio.h>
+#include <stdlib.h>
+#include <stdarg.h>
+
+// function for logging calls
+
+static int is_reset = 0;
+
+void logcall(const char *format, ...)
+{
+    static FILE *file_ptr;
+    if (!is_reset) {
+	file_ptr = fopen("./strace.log", "w");
+	is_reset = 1;
+    }
+    va_list args_ptr;
+    va_start(args_ptr, format);
+    vfprintf(file_ptr, format, args_ptr);
+    va_end(args_ptr);
+}
+
+// wrapper for open()
+
+typedef int (*real_open_t)(const char *, int, ...);
+
+int open(const char *pathname, int flags, ...)
+{
+    int fd;
+    if (flags & (O_CREAT | O_TMPFILE)) {
+	va_list args_ptr;
+	va_start(args_ptr, flags);
+	fd =
+	    ((real_open_t)dlsym(RTLD_NEXT, "open"))
+	    (pathname, flags, va_arg(args_ptr, mode_t));
+	va_end(args_ptr);
+    } else {
+	fd = ((real_open_t)dlsym(RTLD_NEXT, "open"))(pathname, flags);
+    }
+    char *abspath = realpath(pathname, NULL);
+    logcall("open(\"%s\",,) = %d\n", abspath, fd);
+    free(abspath);
+    return fd;
+}
+
+// wrapper for open64()
+
+typedef int (*real_open64_t)(const char *, int, ...);
+
+int open64(const char *pathname, int flags, ...)
+{
+    int fd;
+    if (flags & (O_CREAT | O_TMPFILE)) {
+	va_list args_ptr;
+	va_start(args_ptr, flags);
+	fd =
+	    ((real_open_t)dlsym(RTLD_NEXT, "open64"))
+	    (pathname, flags, va_arg(args_ptr, mode_t));
+	va_end(args_ptr);
+    } else {
+	fd = ((real_open_t)dlsym(RTLD_NEXT, "open64"))(pathname, flags);
+    }
+    char *abspath = realpath(pathname, NULL);
+    logcall("open(\"%s\",,) = %d\n", abspath, fd);
+    free(abspath);
```

Comment:
Here `realpath()` could set `errno` and return `NULL`, which would result in us overwriting the value of `errno` returned to the caller and trying to format `NULL` as `%s` (glibc supports that as an extension, but on many platforms it will segfault).  Even when it works I think it'd be better to show the value of `pathname` instead of something like `<NULL>` in that case.

Reply:
Saving and restoring errno around realpath in the latest commit.